src/com/facebook/buck/android/AndroidInstrumentationTestDescription.java

```diff
@@ -0,0 +1,134 @@
+/*
+ * Copyright 2015-present Facebook, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License. You may obtain
+ * a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.facebook.buck.android;
+
+import com.facebook.buck.model.BuildTarget;
+import com.facebook.buck.model.HasSourceUnderTest;
+import com.facebook.buck.rules.BuildRule;
+import com.facebook.buck.rules.BuildRules;
+import com.facebook.buck.rules.BuildRuleParams;
+import com.facebook.buck.rules.BuildRuleResolver;
+import com.facebook.buck.rules.BuildRuleType;
+import com.facebook.buck.rules.Description;
+import com.facebook.buck.rules.InstallableApk;
+import com.facebook.buck.rules.Label;
+import com.facebook.buck.rules.SourcePathResolver;
+import com.facebook.buck.util.HumanReadableException;
+import com.facebook.infer.annotation.SuppressFieldNotInitialized;
+import com.google.common.base.Optional;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.ImmutableSortedSet;
+
+public class AndroidInstrumentationTestDescription
+    implements Description<AndroidInstrumentationTestDescription.Arg> {
+
+  public static final BuildRuleType TYPE = BuildRuleType.of("android_instrumentation_test");
+
+  private final Optional<Long> testRuleTimeoutMs;
+
+  public AndroidInstrumentationTestDescription(Optional<Long> testRuleTimeoutMs) {
+    this.testRuleTimeoutMs = testRuleTimeoutMs;
+  }
+
+  @Override
+  public BuildRuleType getBuildRuleType() {
+    return TYPE;
+  }
+
+  @Override
+  public Arg createUnpopulatedConstructorArg() {
+    return new Arg();
+  }
+
+  @Override
+  public <A extends Arg> AndroidInstrumentationTest createBuildRule(
+      BuildRuleParams params,
+      BuildRuleResolver resolver,
+      A args) {
+    BuildRule baseApk = resolver.getRule(args.apk);
+    if (!(baseApk instanceof InstallableApk)) {
+      throw new HumanReadableException(
+          "In %s, apk='%s' must be an android_binary() or apk_genrule() but was %s().",
+          params.getBuildTarget(),
+          baseApk.getFullyQualifiedName(),
+          baseApk.getType());
+    }
+
+    if (((InstallableApk) baseApk).getExopackageInfo().isPresent() &&
+        !args.instrumentationApk.isPresent()) {
+      throw new HumanReadableException(
+          "In %s, apk='%s' can not be an exopackage if the instrumentation_apk parameter " +
+          "is not present.",
```

Comment:
I'm not 100% sure why, but when I tried to run the tests with just the exopackage as `apk` and no `instrumentation_apk`, no tests were run.  (Done by commenting out the `instrumentation_apk` line in apps/myapp_test/BUCK and adding `'//test/com/example/activity:activity_test',` in the exopackage `android_binary` rule in apps/myapp/BUCK.  Maybe I'm missing something here (I have very little experience with exopackages), but otherwise I'd prefer to just not allow it instead of having some users confused.